File: lib/python/questionnaire/utility.py
def string_is_valid(string):
    try:
        str(string)
        if len(string) == 0:
            return False

        return True
    except:
        return False


def print_field_error_message():
    print("Please try again, invalid input")


def fetch_user_evaluation(default_message, correct_questions):
    message = ""

    developer_rating = {
        "frontend": 0,
        "backend": 0,
        "devops": 0,
        "mobile": 0
    }

    for question in correct_questions:
        tag = question.tag
        developer_rating[tag] += 1

    frontend = "frontend"
    backend = "backend"
    devops = "devops"
    mobile = "mobile"

    passing_limit = 3
    failure_limit = 2

    is_frontend_developer = developer_rating.get(frontend) > failure_limit and developer_rating.get(
        backend) <= failure_limit and developer_rating.get(devops) <= failure_limit and developer_rating.get(mobile) <= failure_limit

    is_backend_developer = developer_rating.get(frontend) <= failure_limit and developer_rating.get(
        backend) > failure_limit and developer_rating.get(devops) <= failure_limit and developer_rating.get(mobile) <= failure_limit

    is_devops_engineer = developer_rating.get(frontend) < failure_limit and developer_rating.get(
        backend) < failure_limit and developer_rating.get(devops) > failure_limit and developer_rating.get(mobile) < failure_limit

    is_mobile_developer = developer_rating.get(frontend) < failure_limit and developer_rating.get(
        backend) < failure_limit and developer_rating.get(devops) < failure_limit and developer_rating.get(mobile) > failure_limit

    is_full_stack_developer = developer_rating.get(frontend) > failure_limit and developer_rating.get(
        backend) > failure_limit and developer_rating.get(devops) < failure_limit

    is_software_engineer = developer_rating.get(frontend) > failure_limit and developer_rating.get(
        backend) > failure_limit and developer_rating.get(devops) >= failure_limit

    is_all_rounder = developer_rating.get(frontend) >= failure_limit and developer_rating.get(
        backend) >= failure_limit and developer_rating.get(devops) >= failure_limit and developer_rating.get(mobile) >= failure_limit

    # print("frontend developer: " + str(is_frontend_developer))
    # print("backend developer: " + str(is_backend_developer))
    # print("devops developer: " + str(is_devops_engineer))
    # print("mobile developer: " + str(is_mobile_developer))
    # print("fullstack developer: " + str(is_full_stack_developer))
    # print("software engineer: " + str(is_software_engineer))
    # print("all rounder: " + str(is_all_rounder))

    if is_all_rounder:
        message = "WOW! Looks like we got an all rounder, congratulations!!!"
    elif is_software_engineer:
        message = "You're definitely a Software engineer :-)"
    elif is_full_stack_developer:
        message = "Congratulations, Oh Fullstack developer!"
    elif is_devops_engineer:
        message = "You're definitely a DevOps engineer"
    elif is_mobile_developer:
        message = "Looks like you are a mobile developer!"
    elif is_backend_developer:
        message = "Another great backend developer!"
    elif is_frontend_developer:
        message = "Looks like you are a frontend developer, right?"
    else:
        message = default_message

    message += "\n"

    return message

File: src/lib/nodejs/src/questionnaire/utility.js
module.exports.random = {
    randint: (min, max) => Math.floor(Math.random() * (max - min + 1)) + min
}

module.exports.stringIsValid = function (string) {
    try {
        String(string)

        if (string.length === 0)
            return false

        return true
    } catch (e) {
        return false
    }
}

module.exports.printFieldErrorMessage = function () {
    console.log("Please try again, invalid input")
}

module.exports.fetchUserEvaluation = function (defaultMessage, correctQuestions) {
    let message

    const developerRating = {
        frontend: 0,
        backend: 0,
        devops: 0,
        mobile: 0
    }

    for (const question of correctQuestions) {
        const tag = question.tag
        developerRating[tag] += 1
    }

    const frontend = "frontend"
    const backend = "backend"
    const devops = "devops"
    const mobile = "mobile"

    passingLimit = 3
    failureLimit = 2

    isFrontendDeveloper = developerRating[frontend] > failureLimit && developerRating[backend] <= failureLimit && developerRating[devops] <= failureLimit && developerRating[mobile] <= failureLimit

    isBackendDeveloper = developerRating[frontend] <= failureLimit && developerRating[backend] > failureLimit && developerRating[devops] <= failureLimit && developerRating[mobile] <= failureLimit

    isDevopsEngineer = developerRating[frontend] < failureLimit && developerRating[backend] < failureLimit && developerRating[devops] > failureLimit && developerRating[mobile] < failureLimit

    isMobileDeveloper = developerRating[frontend] < failureLimit && developerRating[backend] < failureLimit && developerRating[devops] < failureLimit && developerRating[mobile] > failureLimit

    isFullstackDeveloper = developerRating[frontend] > failureLimit && developerRating[backend] > failureLimit && developerRating[devops] < failureLimit

    isSoftwareEngineer = developerRating[frontend] > failureLimit && developerRating[backend] > failureLimit && developerRating[devops] >= failureLimit

    isAllRounder = developerRating[frontend] >= failureLimit && developerRating[backend] >= failureLimit && developerRating[devops] >= failureLimit && developerRating[mobile] >= failureLimit

    // console.log("frontend developer: " + isFrontendDeveloper)
    // console.log("backend developer: " + isBackendDeveloper)
    // console.log("devops developer: " + isDevopsEngineer)
    // console.log("mobile developer: " + isMobileDeveloper)
    // console.log("fullstack developer: " + isFullstackDeveloper)
    // console.log("software engineer: " + isSoftwareEngineer)
    // console.log("all rounder: " + isAllRounder)

    if (isAllRounder)
        message = "WOW! Looks like we got an all rounder, congratulations!!!"
    else if (isSoftwareEngineer)
        message = "You're definitely a Software engineer :-)"
    else if (isFullstackDeveloper)
        message = "Congratulations, Oh Fullstack developer!"
    else if (isDevopsEngineer)
        message = "You're definitely a DevOps engineer"
    else if (isMobileDeveloper)
        message = "Looks like you are a mobile developer!"
    else if (isBackendDeveloper)
        message = "Another great backend developer!"
    else if (isFrontendDeveloper)
        message = "Looks like you are a frontend developer, right?"
    else
        message = defaultMessage

    message += "\n"

    return message
}